File: python/match_av.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import pandas as pd
import numpy as np
from splink.duckdb.linker import DuckDBLinker
from splink.duckdb.blocking_rule_library import block_on
import altair as alt
import splink.duckdb.comparison_library as cl
import splink.duckdb.comparison_template_library as ctl ## Hey they also have template functions
import string
from time import time
import pickle

from python.match_settings import match_vars, match_settings, block_vars, comp_vars

# Task 1: Match CT40_output2 to CT40_aug
# Task 2: Match CT40_output2 to CT99_aut
# Task 3: Match CT40_output3 to CT40_aug
# Task 4: Match CT40_output3 to CT99_aut

#df1 = pd.read('CT40_output3')
df1_targs = ['CT40_output3', 'CT40_output2']
df2_targs = ['CT99_aug', 'CT40_aug']


for df1t in df1_targs:
    for df2t in df2_targs:

        df1 = pd.read_csv(df1t+'.csv')
        df2 = pd.read_csv('./data/'+df2t+'.csv')

        ## Add subset vars.
        df1['ch2_first_name'] = [x[:2] if type(x)==str else x for x in df1['first_name']]
        df2['ch2_first_name'] = [x[:2] if type(x)==str else x for x in df2['first_name']]

        df1['first_initial'] = [x[:1] if type(x)==str else x for x in df1['first_name']]
        df2['first_initial'] = [x[:1] if type(x)==str else x for x in df2['first_name']]

        df1['ch4_last_name'] = [x[:4] if type(x)==str else x for x in df1['last_name']]
        df2['ch4_last_name'] = [x[:4] if type(x)==str else x for x in df2['last_name']]

        df1['zip3'] = [x[:3] if type(x)==str else x for x in df1['mailing_address_zipcode']]
        df2['zip3'] = [x[:3] if type(x)==str else x for x in df2['mailing_address_zipcode']]

        df1['ssn4'] = [x[-4:] if type(x)==str else x for x in df1['ssn']]
        df2['ssn4'] = [x[-4:] if type(x)==str else x for x in df2['ssn']]

        ## Keep only subset cols
        df1['unique_id'] = df1.index
        df2['unique_id'] = df2.index

        df1['age'] = df1['age'].astype(str)
        df2['age'] = df2['age'].astype(str)

        precs = {}
        recls = {}
        exec_time = {}
        for ms in string.ascii_lowercase[:22]:
            for i in range(10):
                print('------------------')
            print(ms)
            for i in range(10):
                print('------------------')

            tstart = time()

            keepcols = ['unique_id'] + match_vars[ms]

            df1_sub = df1[keepcols]
            df2_sub = df2[keepcols]

            if ms=='b':
                mdf = pd.merge(df1_sub, df2_sub, on='ssn', how = 'right')
                isnan = np.array([np.isnan(x) if type(x)==float else False for x in mdf['ssn']])
                mdf = mdf.loc[~isnan,:]
                mdf = mdf.rename({'unique_id_x':'unique_id_l', 'unique_id_y':'unique_id_r'}, axis = 1)
            else:
                #block_rule = block_on(block_vars[ms])

                settings = match_settings[ms]

                linker = DuckDBLinker([df1_sub, df2_sub], settings)

                ### Estimate parameters
                drs = "AND".join([' (l.'+v+' = r.'+v+') ' for v in match_vars[ms]])
                deterministic_rules = [drs]

                linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)
                linker.estimate_u_using_random_sampling(max_pairs=1e6)

                for v in comp_vars[ms]:
                    training_blocking_rule = block_on([v])
                    training_session_fname_sname = linker.estimate_parameters_using_expectation_maximisation(training_blocking_rule)

                ### Do actual matching.
                df_predictions = linker.predict(threshold_match_probability=0.1)
                mdf = df_predictions.as_pandas_dataframe()

            ### Evaluate
            df1_key = df1[['simulant_id','unique_id']]
            df1_key = df1_key.rename(dict([(v,v+'_l') for v in df1_key]), axis = 1)
            mdf = mdf.merge(df1_key, on = 'unique_id_l')

            df2_key = df2[['simulant_id','unique_id']]
            df2_key = df2_key.rename(dict([(v,v+'_r') for v in df2_key]), axis = 1)
            mdf = mdf.merge(df2_key, on = 'unique_id_r')

            num_correctly_matched = np.sum(mdf['simulant_id_l']==mdf['simulant_id_r'])
            num_matched = mdf.shape[0]
            num_incorrectly_matched = np.sum(mdf['simulant_id_l']!=mdf['simulant_id_r'])
            common_records = len(set(df1['simulant_id']).intersection(df2['simulant_id']))

            recall = num_correctly_matched/common_records
            precision = num_correctly_matched/num_matched

            precs[ms] = precision
            recls[ms] = recall
            print('------------------------')
            print(ms)
            print("Precision: %f"%precision)
            print("Recall: %f"%recall)
            print('------------------------')

            exec_time[ms] = time() - tstart

        #with open("pickles/match_dmp.pkl", 'wb') as f:
        #    pickle.dump([precs, recls, exec_time], f)
        with open("pickles/"+df1t+"_to_"+df2t+".pkl", 'wb') as f:
            pickle.dump([precs, recls, exec_time], f)

File: R/match_av.R
#!/usr/bin/Rscript
#  match_av.R Author "Nathan Wycoff <nathanbrwycoff@gmail.com>" Date 07.04.2024

library(fastLink)
source("R/match_settings.R")

#mode <- 'default'
mode <- 'splink'

df1_targs <- c('CT40_output3', 'CT40_output2')
df2_targs <- c('CT99_aug', 'CT40_aug')

for (df1t in df1_targs) {
    for (df2t in df2_targs) {
        df1 <- read.csv(paste(df1t,'.csv', sep =''))
        df2 <- read.csv(paste('./data/',df2t,'.csv', sep =''))

        ## Add subset vars.
        df1$ch2_first_name <- substr(df1$first_name, 1, 2)
        df2$ch2_first_name <- substr(df2$first_name, 1, 2)

        df1$first_initial <- substr(df1$first_name, 1, 1)
        df2$first_initial <- substr(df2$first_name, 1, 1)

        df1$ch4_last_name <- substr(df1$last_name, 1, 4)
        df2$ch4_last_name <- substr(df2$last_name, 1, 4)

        df1$zip3 <- substr(df1$mailing_address_zipcode, 1, 3)
        df2$zip3 <- substr(df2$mailing_address_zipcode, 1, 3)

        df1$ssn4 <- substr(df1$ssn, 11-3,11)
        df2$ssn4  <- substr(df2$ssn, 11-3,11)

        ## Loop over match passes here.
        tta <- Sys.time()
        precs <- rep(NA, 22)
        recls <- rep(NA, 22)
        times <- rep(NA, 22)
        names(precs) <- names(recls) <- names(times) <- letters[1:22]
        for (ms in letters[1:22]) {
            print(ms)
            tt <- Sys.time()
            #keepcols <- c("first_name", "last_name")

            ## Blocking
            if (length(block_vars[[ms]])>0) {
                capture.output(blocks <- blockData(df1, df2, varnames = block_vars[[ms]]))
            } else {
                blocks <- list(list(dfA.inds = 1:nrow(df1), dfB.inds = 1:nrow(df2)))
            }
            print(length(blocks))

            ## Fuzzy match settings
            if (comp_type[[ms]]=='fuzzy') {
                if (mode=='default') {
                    options <- list()
                } else if (mode=='splink') {
                    options <- list(cut.a=0.9, cut.p = 0.8)
                }
            } else if (comp_type[[ms]]=='exact') {
                options <- list('cut.a' = 1., 'cut.p' = 1, 'cut.a.num' = 0., 'cut.p.num' = 0.)
            } else {
                stop("Bad comptype")
            }
            options$n.cores = 40

            if (length(blocks)>1) {
                preoptions <- options
                preoptions$dfA <- df1
                preoptions$dfB <- df2
                preoptions$varnames = comp_vars[[ms]]
                capture.output(pre_ret <- do.call(fastLink, preoptions))
            }

            ## Actual Linking
            m <- matrix(NA,nrow=0,ncol=2)
            colnames(m) <-c("inds.a", "inds.b") 
            minrow <- 3
            bi <- 0
            for (block in blocks) {
                bi <- bi + 1
                print(bi)
                #isbad <- (bi==174) & ms=='e' & df1t == 'CT40_output2' & df2t == 'CT40_aug'
                #isbad <- (bi==1) & ms=='e' & df1t == 'CT40_output2' & df2t == 'CT40_aug'
                options$dfA = df1[block$dfA.inds,]
                options$dfB = df2[block$dfB.inds,]
                options$varnames = comp_vars[[ms]]
                #isbad <- ((dim(options$dfA)[1]) == 0 || dim(options$dfB)[1]==0)
                isbad <- dim(options$dfA)[1] <= minrow || dim(options$dfB)[1] <= minrow
                print(dim(options$dfA))
                print(dim(options$dfB))
                print(isbad)
                if (isbad) {
                    print("Skipped")
                } else {
                    #capture.output(ret <- fastLink(dfA = df1[block$dfA.inds,], dfB = df2[block$dfB.inds,], varnames = comp_vars[[ms]], n.cores = 40)) 
                    #capture.output(ret <- do.call(fastLink, options))
                    if (length(blocks)>1) {
                        options$em.obj <- pre_ret$EM
                    }
                    capture.output(ret <- do.call(fastLink, options))
                    ret$matches$inds.a <- block$dfA.inds[ret$matches$inds.a]
                    ret$matches$inds.b <- block$dfB.inds[ret$matches$inds.b]
                    m <- rbind(m, ret$matches)
                }
            }

            ## Extract match information and store.
            sid1 <- df1[m[,1],'simulant_id']
            sid2 <- df2[m[,2],'simulant_id']

            num_correctly_matched <- sum(sid1==sid2)
            num_matched <- nrow(m)
            num_incorrectly_matched <- sum(sid1!=sid2)
            common_records <- length(intersect(df1$simulant_id, df2$simulant_id))

            recall <- num_correctly_matched/common_records
            precision <- num_correctly_matched/num_matched

            print(recall)
            print(precision)

            print("pr:")
            precs[ms] = precision
            recls[ms] = recall
            times[ms] <- as.numeric(Sys.time()-tt, units = 'secs')
            print("time:")
            print(Sys.time()-tt
        ) 
        }

        resdf <- data.frame(Precision = precs, Recall = recls, Time = times)
        write.csv(resdf, file = paste('sim_out/fastLink_', df1t, '_', df2t, '_',mode,'.csv', sep=''))

        print("overall:")
        print(Sys.time() - tta)
    }
}
